be/src/exec/pipeline/stream_pipeline_driver.cpp

```diff
@@ -0,0 +1,293 @@
+                return _state;
+            }
+        }
+
+        // no chunk moved in current round means that the driver is blocked.
+        // should yield means that the CPU core is occupied the driver for a
+        // very long time so that the driver should switch off the core and
+        // give chance for another ready driver to run.
+        if (num_chunks_moved == 0 || should_yield) {
+            if (is_precondition_block()) {
+                set_driver_state(DriverState::PRECONDITION_BLOCK);
+                COUNTER_UPDATE(_block_by_precondition_counter, 1);
+            } else if (!sink_operator()->is_finished() && !sink_operator()->need_input()) {
+                set_driver_state(DriverState::OUTPUT_FULL);
+                COUNTER_UPDATE(_block_by_output_full_counter, 1);
+            } else if (!source_operator()->is_finished() && !source_operator()->has_output()) {
+                set_driver_state(DriverState::INPUT_EMPTY);
+                COUNTER_UPDATE(_block_by_input_empty_counter, 1);
+            } else {
+                set_driver_state(DriverState::READY);
+            }
+            return _state;
+        }
+    }
+}
+
+Status StreamPipelineDriver::_mark_operator_epoch_finishing(OperatorPtr& op, RuntimeState* state) {
+    auto& op_state = _operator_stages[op->get_id()];
+    if (op_state >= OperatorStage::EPOCH_FINISHING) {
+        return Status::OK();
+    }
+
+    VLOG_ROW << strings::Substitute("[Driver] epoch_finishing operator [fragment_id=$0] [driver=$1] [operator=$2]",
+                                    print_id(state->fragment_instance_id()), to_readable_string(), op->get_name());
+    {
+        SCOPED_TIMER(op->_finishing_timer);
+        op_state = OperatorStage::EPOCH_FINISHING;
+        QUERY_TRACE_SCOPED(op->get_name(), "set_epoch_finishing");
+        return op->set_epoch_finishing(state);
+    }
+}
+
+Status StreamPipelineDriver::_mark_operator_epoch_finished(OperatorPtr& op, RuntimeState* state) {
+    auto& op_state = _operator_stages[op->get_id()];
+    if (op_state >= OperatorStage::EPOCH_FINISHED) {
+        return Status::OK();
+    }
+
+    VLOG_ROW << strings::Substitute("[Driver] epoch_finished operator [fragment_id=$0] [driver=$1] [operator=$2]",
+                                    print_id(state->fragment_instance_id()), to_readable_string(), op->get_name());
+    {
+        SCOPED_TIMER(op->_finishing_timer);
+        op_state = OperatorStage::EPOCH_FINISHED;
+        QUERY_TRACE_SCOPED(op->get_name(), "set_epoch_finished");
+        return op->set_epoch_finished(state);
+    }
+}
+
+Status StreamPipelineDriver::reset_epoch(RuntimeState* runtime_state) {
+    for (auto& op : _operators) {
+        RETURN_IF_ERROR(op->reset_epoch(runtime_state));
+        _operator_stages[op->get_id()] = OperatorStage::PREPARED;
+    }
+    _first_epoch_unfinished = 0;
+    _num_epoch_finished_drivers = _fragment_ctx->pipelines().size();
```

Comment:
Why `_num_epoch_finished_drivers` is equal to the number of pipelines not that of drivers?

Reply:
fixed.